popover: Fix user pop over not showing when clicked in channels /group settings

PieterCK opened the pull request:
## :bookmark_tabs: Overview
This PR aims to continue where #28820 has left off. The changes introduced in my PR are primarily fixes addressing the comments left by reviewer in the original PR.

Details of the original PR by Ecxtacy: 
- Added css fix for hovering effect on the X icon in user pills.
- Added an optimization refactor to `removePill` function of `input_pill.ts` module as discussed in the [CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Input.20pill.20code.20optimization).
- Fixed user pill related components to enable popover effect for user and user group pills in channel and group settings.

New changes in this PR is detailed in the PR comments I've highlighted in the code.  

Fixes: #28687

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>
Popovers on user and group pills in channel and group menu:
</summary>

Relevant commits for this change:
- `input_pill: Pass user_id and group_id to render pill.`
- `settings: Show user_card_popover when pill is clicked.`
- `popovers: Refactor user_group_popover to work with pills.`
- `settings: Show user_group_popover when pill is clicked.`

Popovers on Channel/Stream settings:

| **Before** | **After** |
|---|---|
|![image](https://github.com/zulip/zulip/assets/108744694/80de23bf-701a-4812-b453-91dd5a2d2681)|![image](https://github.com/zulip/zulip/assets/108744694/22ffb0b8-bccb-4ed1-a832-d65fd3a98a27)![image](https://github.com/zulip/zulip/assets/108744694/9977552f-cf09-48c8-a5d1-cbf9de5fde25)|

Popovers on User Groups Settings:

| **Before** | **After** |
|---|---|
|![image](https://github.com/zulip/zulip/assets/108744694/34b76dcb-7b1b-4992-9268-2fe7194eb64b)|![image](https://github.com/zulip/zulip/assets/108744694/c10e3ed7-7606-4caf-9727-4939546157ce)![image](https://github.com/zulip/zulip/assets/108744694/484dd784-eec6-4a98-8aae-ebf1638c1ab8)|
</details>


<details>
<summary> X button hover effect:
</summary>

Relevant commit for this change: `css: Show 'x' button hover styling when hovered over 'x'.`

Before:
X button focuses even only by hovering over the pill and not the button itself.

[Screencast from 04-05-24 16:06:44.webm](https://github.com/zulip/zulip/assets/108744694/f54f928b-18ed-4682-b3b1-bb0754c0d9e0)

After:
X button only focuses if the cursor is on top of it.
![image](https://github.com/zulip/zulip/assets/108744694/808b2797-fd70-4bbd-87a9-0933a6e92790)
![image](https://github.com/zulip/zulip/assets/108744694/304d6611-c532-46e9-a143-5ae3d8bba12e)

[Screencast from 04-05-24 14:38:21.webm](https://github.com/zulip/zulip/assets/108744694/f8bb25c3-38cb-486e-be1a-09736551247d)
</details>

<details>
<summary>Popover closing behavior:
</summary>

Relevant commit for this change: `hashchange: Fix popover not disappearing on overlay hash URL.`

Before: 
Clicking back button in browser doesn't hide popovers

[Screencast from 04-05-24 15:40:48.webm](https://github.com/zulip/zulip/assets/108744694/96971ce5-74ca-4c98-b958-ab29e885676c)

After:

[Screencast from 04-05-24 15:57:19.webm](https://github.com/zulip/zulip/assets/108744694/239b453d-eab9-47b0-abea-d2e78d53ff6d)

[Screencast from 04-05-24 15:57:53.webm](https://github.com/zulip/zulip/assets/108744694/bfd11f44-070d-437a-b107-4cb8ce3cd8a0)

</details>

<details>
<summary>Impacted functions, no regression found:
</summary>

`toggle_user_group_info_popover()` in `user_group_popover.js` still function on mentions in messages:
![image](https://github.com/zulip/zulip/assets/108744694/d402ba6f-087d-46c6-ba49-4ed98df51055)

other user pills in group & channel setting:
![image](https://github.com/zulip/zulip/assets/108744694/dffe1558-094e-40b4-99f1-92e72aa56f2f)

[Screencast from 03-05-24 19:58:05.webm](https://github.com/zulip/zulip/assets/108744694/cb067179-10b5-4188-8fc2-18c02961a994)
</details>

### :factory_worker:  Concerns & Issue Encountered
- inconsistency in `user_group_pill.ts`. I've pointed out the detail [here](https://github.com/zulip/zulip/pull/29898/files#r1588885120)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented:
umm, hello @ecxtacy. I think I did an oopsie by mentioning(@) you in the PR commits. I think you'll get notifications eveytime I push something here. I'm not sure whether that stays true even if I've changed the commit descriptions to not mention you. Sorry for that, let me know if you need help turning it off :sweat_smile:

alya commented:
Thanks! To prepare this PR for review, you'll need to complete the self-review checklist, and reply to @sahil839 's comments on https://github.com/zulip/zulip/pull/28820, which had some questions. You can do so by commenting on this PR, including the original comment and your response.

PieterCK commented on web/src/input_pill.ts:
```diff
@@ -19,6 +19,8 @@ export type InputPillItem<T> = {
     status_emoji_info?: EmojiRenderingDetails & {emoji_alt_code?: boolean}; // TODO: Move this in user_status.js
     should_add_guest_user_indicator?: boolean;
+    user_id?: number;
+    id?: number;
```
Original Review:
> sahil: Why not group_id here instead of id?

Hi Sahil, it's because `user_group_pill.ts` used `id` instead if `group_id`. I also agree that it be `group_id`, consistent with `user_id` instead of just `id`. I can add it as a prep commit if you think we should make the change

_relevant code in_ `user_group_pill.ts`: 
![image](https://github.com/zulip/zulip/assets/108744694/ab64ddee-eb40-4838-9e67-4f21955a1e58)

PieterCK commented on web/src/hashchange.js:
```diff
@@ -471,6 +471,8 @@ function hashchanged(from_reload, e) {
         return undefined;
     }
 
+    popovers.hide_all();
```
Original Comment on `web/src/user_card_popover.js` line 367:
> sahil: I guess this is for hiding the popover when pressing browser back button. I think we should update the code in hashchange.js. You can see we have hide_all calls, which are used to hide all popovers, etc. there.

I've deleted the new event listeners for closing the popovers and instead refactored `hashchange.js` to achieve the same effect.

PieterCK commented on web/templates/input_pill.hbs:
```diff
@@ -1,4 +1,4 @@
-<div class='pill {{#if deactivated}} deactivated-pill {{/if}}' tabindex=0>
+<div class='pill {{#if deactivated}} deactivated-pill {{/if}}'{{#if user_id}}data-user-id="{{user_id}}"{{/if}}{{#if group_id}}data-group-id="{{group_id}}"{{/if}} tabindex=0>
```
Original Review:
> sahil: I think we should name these attributes to data-user-id and data-group-id.

Changed it! :ok_hand:

PieterCK commented on web/src/user_card_popover.js:
```diff
@@ -785,13 +785,17 @@ function register_click_handlers() {
-    });
+    $("body").on(
+        "click",
+        ".view_user_profile, .pill-container.person_picker > .pill[data-user-id]",
```
In the original PR, there's new separate function responsible for adding event listener to the pill on stream and group setting. In the original PR, it was on line 885 in `web/src/user_card_popover.js`. However, it introduced a new bug because the pill container targeted by the event listener this comment is on is also activated and ended up popping up double popover when clicked. So, I added a second selector to this function instead.

_double popover bug:_
![Screenshot from 2024-04-30 20-46-58](https://github.com/zulip/zulip/assets/108744694/4734f461-5c64-4ab3-a31a-f05267cc8152)

PieterCK commented on web/src/user_group_popover.ts:
```diff
@@ -130,6 +140,17 @@ export function register_click_handlers(): void {
     });
+
+    // Show the user_group_popover when pill clicked in subscriber settings.
+    $("body").on("click", ".pill-container.person_picker > .pill[data-group-id]", (e) => {
```
Narrowed down the JQuery selector for the pills. 
Original: `.pill[group-id]`

PieterCK commented on web/src/user_group_popover.ts:
```diff
@@ -53,17 +53,19 @@ export function handle_keyboard(key: string): void {
 export function toggle_user_group_info_popover(
     element: ReferenceElement,
-    message_id: number,
+    message_id: number | undefined,
```
Original review:
> sahil: Do we need this change? We would anyway pass element and that can be used to get the group object and message_id can just be passed as undefined.

I agree, refactored the function to accept user group pill as element.

PieterCK commented:
Update: Edited the PR description, added new changes to the PR addressing reviews from the original PR and highlighted those changes through PR comments.

PieterCK commented:
> Thanks! To prepare this PR for review, you'll need to complete the self-review checklist, and reply to @sahil839 's comments on #28820, which had some questions. You can do so by commenting on this PR, including the original comment and your response.

Sorry for notifying a sloppy PR, I hope it's clearer now. Thank you as always!

alya commented:
Awesome, thanks! @sahil839 could you please take a look next?

I can manually test too once you're happy with the code.

sahil839 reviewed (commented):
Posted few comments. I think you would want to update the commit messages to mention yourself as [co-author](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors), as the commits are still authored by @ecxtacy.

PieterCK commented on web/src/hashchange.js:
```diff
@@ -487,6 +487,11 @@ function hashchanged(from_reload, e) {
 
+    // Hide popovers when user is switching tab / section the settings menu
+    if (hash_parser.is_in_specified_hash_category(["stream", "channels", "groups"])) {
+        popovers.hide_all();
```
yeah, initially I thought of implementing it that way. But I'm not sure about the reason why (if there was any) `popovers.hide_all()` was put after the `is_overlay_hash` check in the first place, so I opted implementing an allowlist kind of check just to be conservative. 
Though when testing locally I couldn't find any regression when it was only just `popovers.hide_all()` before the overlay check with out the extra `is_in_specified_hash_category` check. For now I'll revert it to just `popovers.hide_all()` since I couldn't think of any breaking regression, so its probably fine to do it that way. :+1:

sahil839 replied:
It is good to check if there was any specific reason for the code change by using `git blame`. This line was originally added in 770a7896515b4 and it does not mention any specific reason to call this after the overlay change code.

Anyways, I think it is fine to just hide the popovers in case the view changes, which is what happens on changing the hash, so this change should be fine. We can see if there is any other suggestion/feedback when the PR is reviewed by Tim.

sahil839 reviewed (commented):
Posted a few more comments. Looks good otherwise.

PieterCK commented on web/src/hash_parser.ts:
```diff
@@ -111,6 +111,11 @@ export function is_subscribers_section_opened_for_stream(): boolean {
     return hash_components[3] === "subscribers";
 }
 
+export function is_in_specified_hash_category(hash_categories: string[]): boolean {
```
Yeah, I separated it into its own commit just in case it's not needed. It can be left out if the other commits are ready for to be merged

sahil839 replied:
In that case, I think you can keep this commit at the end.

sbansal1999 commented:
Yeah, looks like some issue during provisioning. You can re-run the tests by force-pushing your commits again.

Do it by amending your last commit's message using `git commit --amend` and then saving the commit message without making any changes. After that you can do a force push, which will trigger the tests again.

PieterCK commented:
Aha, passed. Thanks for the tip!

sahil839 commented:
Code looks good. @alya this is ready for testing.

A couple of points -
- We still change the color of pill to green on clicking on them.
- This PR also adds code to show the pills when clicking on pills in a person type custom profile field, like `Mentor` field in "Profile" panel or manage user form in development environment due to elements using same class, but I guess that should be fine.

sahil839 commented:
Also, there is a PR #24334, which adds support for opening the popover for all the pills. This PR can be followed by completing that PR.

alya commented:
@PieterCK Please bump for my review once conflicts are resolved.

PieterCK commented:
Resolved merge conflict. @alya this is ready for testing

alya commented:
Works for me in manual testing!

alya commented:
@timabbott This PR addresses @sahil839 's feedback on an earlier PR.

timabbott commented:
Merged, thanks @PieterCK and @sahil839!
